src/unix/linux/android/Matoya.java

```diff
@@ -206,16 +205,24 @@ static boolean isMouseEvent(InputEvent event) {
 	}
 
 	static boolean isGamepadEvent(InputEvent event) {
-		return
-			(event.getSource() & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD ||
-			(event.getSource() & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK ||
-			(event.getSource() & InputDevice.SOURCE_DPAD) == InputDevice.SOURCE_DPAD;
+		return event.getDevice().getControllerNumber() != 0;
+	}
+
+	static boolean hasAxisTriggers(InputDevice device) {
+		for (InputDevice.MotionRange range : device.getMotionRanges()) {
+			if (range.getAxis() == MotionEvent.AXIS_LTRIGGER || range.getAxis() == MotionEvent.AXIS_RTRIGGER)
+				return true;
+		}
+
+		return false;
 	}
 
 	boolean keyEvent(int keyCode, KeyEvent event, boolean down) {
 		// Button events fire here (sometimes dpad)
-		if (isGamepadEvent(event))
-			app_button(event.getDeviceId(), down, keyCode);
+		if (isGamepadEvent(event)) {
+			boolean axis_triggers = hasAxisTriggers(event.getDevice());
+			app_button(event.getDeviceId(), down, keyCode, axis_triggers);
+		}
```

Comment:
Shouldn't this return, to avoid sending as both gamepad _and_ keyboard

Reply:
It... might? I actually questioned myself about it. Returning means that uncommon devices that have an integrated keyboard (e.g. [maybe this](https://m.media-amazon.com/images/I/6130ZDH+18L._AC_UF894,1000_QL80_.jpg)) would only work as gamepad, while they actually can write letters. But then I can't test such device, and I'm not even sure it would act as a single device, so it might be safer to return... I'm not that opinionated, I'll do as you prefer.

Reply:
It can be left as-is. That way we'll hear if there are issues at least